pkg/reduce/pnf/ordered.go

```diff
@@ -0,0 +1,243 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pnf
+
+import (
+	"container/list"
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/watermark/wmb"
+	"go.uber.org/zap"
+
+	"github.com/numaproj/numaflow/pkg/forward"
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/metrics"
+	"github.com/numaproj/numaflow/pkg/reduce/applier"
+	"github.com/numaproj/numaflow/pkg/reduce/pbq"
+	"github.com/numaproj/numaflow/pkg/reduce/pbq/partition"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	"github.com/numaproj/numaflow/pkg/watermark/publish"
+)
+
+var retryDelay = 1 * time.Second
+
+// ForwardTask wraps the `processAndForward`.
+type ForwardTask struct {
+	// doneCh is used to notify when the ForwardTask has been completed.
+	doneCh chan struct{}
+	pf     *processAndForward
+}
+
+// OrderedProcessor orders the forwarding of the result of the execution of the tasks, even though the tasks itself are
+// run concurrently in an out of ordered fashion.
+type OrderedProcessor struct {
+	vertexName    string
+	pipelineName  string
+	vertexReplica int32
+	sync.RWMutex
+	taskDone            chan struct{}
+	taskQueue           *list.List
+	pbqManager          *pbq.Manager
+	udf                 applier.ReduceApplier
+	toBuffers           map[string]isb.BufferWriter
+	whereToDecider      forward.ToWhichStepDecider
+	watermarkPublishers map[string]publish.Publisher
+	idleManager         *wmb.IdleManager
+	log                 *zap.SugaredLogger
+}
+
+// NewOrderedProcessor returns an OrderedProcessor.
+func NewOrderedProcessor(ctx context.Context,
+	vertexInstance *dfv1.VertexInstance,
+	udf applier.ReduceApplier,
+	toBuffers map[string]isb.BufferWriter,
+	pbqManager *pbq.Manager,
+	whereToDecider forward.ToWhichStepDecider,
+	watermarkPublishers map[string]publish.Publisher,
+	idleManager *wmb.IdleManager) *OrderedProcessor {
+	of := &OrderedProcessor{
+		vertexName:          vertexInstance.Vertex.Spec.Name,
+		pipelineName:        vertexInstance.Vertex.Spec.PipelineName,
+		vertexReplica:       vertexInstance.Replica,
+		taskDone:            make(chan struct{}),
+		taskQueue:           list.New(),
+		pbqManager:          pbqManager,
+		udf:                 udf,
+		toBuffers:           toBuffers,
+		whereToDecider:      whereToDecider,
+		watermarkPublishers: watermarkPublishers,
+		idleManager:         idleManager,
+		log:                 logging.FromContext(ctx),
+	}
+
+	go of.forward(ctx)
+
+	return of
+}
+
+func (op *OrderedProcessor) InsertTask(t *ForwardTask) {
+	op.Lock()
+	defer op.Unlock()
+	op.taskQueue.PushBack(t)
+}
+
+// SchedulePnF creates and schedules the PnF routine.
+func (op *OrderedProcessor) SchedulePnF(
+	ctx context.Context,
+	partitionID partition.ID) *ForwardTask {
+
+	pbq := op.pbqManager.GetPBQ(partitionID)
```

Comment:
```suggestion
	_pbq := op.pbqManager.GetPBQ(partitionID)
```

`pbq` will collide with package name